plutus-core/plutus-core/src/Bitwise.hs

```diff
@@ -0,0 +1,569 @@
+    go dst src offset lim
+      | offset == lim = pure ()
+      | otherwise = do
+          let offset' = offset * sizeOf (undefined :: a)
+          block :: a <- peek . plusPtr src $ offset'
+          poke (plusPtr dst offset') . complement $ block
+          go dst src (offset + 1) lim
+
+-- Helpers
+
+isPositivePowerOf256 :: Word8 -> ByteString -> Bool
+isPositivePowerOf256 w8 bs = w8 == 0x80 && BS.all (== zeroBits) bs
+
+-- We compute the read similarly to how we determine the change when we write.
+-- The only difference is that the mask is used on the input to read it, rather
+-- than to modify anything.
+dangerousRead :: ByteString -> Integer -> Bool
+dangerousRead bs i =
+  let (bigOffset, smallOffset) = i `quotRem` 8
+      bigIx = BS.length bs - fromIntegral bigOffset - 1
+      mask = bit (fromIntegral smallOffset) in
+    case mask .&. BS.index bs bigIx of
+      0 -> False
+      _ -> True
+
+-- Important note: this function is only safe under the following conditions:
+--
+-- * The IO used in the function argument only performs memory allocations using
+-- malloc, as well as reads and writes via the Storable interface;
+-- * The pointer argument is only read from, not written to;
+-- * The result of the function argument points to freshly-allocated, malloced
+-- memory; and
+-- * The result of the function argument points to memory whose length matches
+-- that of the input ByteString (in bytes)
+--
+-- Even though a ByteString is represented as Ptr CChar, we can ignore sign (we
+-- only treat them as binary data anyway), and on POSIX platforms (which GHC
+-- silently assumes, even on Windows), CChar _must_ be exactly a byte. Thus, we
+-- allow working over a pointer to Word8 instead, to avoid issues with signs.
+overPtrLen :: ByteString -> (Ptr Word8 -> Int -> IO (Ptr Word8)) -> ByteString
+overPtrLen bs f = unsafeDupablePerformIO . unsafeUseAsCStringLen bs $
+  \(ptr, len) -> f (castPtr ptr) len >>= \p ->
+    unsafePackMallocCStringLen (castPtr p, len)
+
+-- Two's complement in a signed, unbounded representation is somewhat
+-- problematic: in our particular case, we hit this issue on exact powers of
+-- 256. This issue stems from such values (or rather, the ByteString
+-- representations of such) having a two's complement identical to themselves,
+-- as well as a trailing 1. This means that we can't distinguish between a
+-- _negative_ and a _positive_ power from representation alone, and must default
+-- one way or the other.
+--
+-- Thus, when we want to produce a negative representation, we have to ensure
+-- that we 'mark' the result in a way that ensures we can detect that it was
+-- negative. We do this by padding with trailing ones.
+twosCompToNegative :: ByteString -> ByteString
+twosCompToNegative bs = case twosComp bs of
+  bs' -> if bs == bs'
+         then BS.cons (complement zeroBits) bs'
+         else bs'
+
+-- If we're taking a two's complement to produce a positive representation,
```

Comment:
The twos complement stuff is a bit spread out over the file, perhaps we could do with a single Note that explains the various different things we're doing in one place.